File: src/test/java/com/hileco/cortex/instructions/ProgramBuilderTest.java
package com.hileco.cortex.instructions;

import com.hileco.cortex.context.ProcessContext;
import com.hileco.cortex.context.Program;
import com.hileco.cortex.context.ProgramContext;
import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;

public class ProgramBuilderTest {

    @Test
    public void testJump() throws ProgramException {
        ProgramBuilder builder = new ProgramBuilder();
        builder.PUSH(new byte[]{123});
        builder.PUSH(new byte[]{123});
        builder.EQUALS();
        builder.PUSH(BigInteger.valueOf(10L).toByteArray());
        builder.JUMP_IF();
        builder.NOOP();
        builder.NOOP();
        builder.NOOP();
        builder.NOOP();
        builder.NOOP();
        builder.JUMP_DESTINATION();
        Program program = builder.build();
        ProgramContext programContext = new ProgramContext(program);
        ProcessContext processContext = new ProcessContext(programContext);
        ProgramRunner programRunner = new ProgramRunner(processContext);
        programRunner.run();
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionPosition());
        Assert.assertEquals(6, programContext.getInstructionsExecuted());
    }

    @Test
    public void testNoJump() throws ProgramException {
        ProgramBuilder builder = new ProgramBuilder();
        builder.PUSH(new byte[]{123});
        builder.PUSH(new byte[]{124});
        builder.EQUALS();
        builder.PUSH(BigInteger.valueOf(10L).toByteArray());
        builder.JUMP_IF();
        builder.NOOP();
        builder.NOOP();
        builder.NOOP();
        builder.NOOP();
        builder.NOOP();
        builder.JUMP_DESTINATION();
        Program program = builder
                .build();
        ProgramContext programContext = new ProgramContext(program);
        ProcessContext processContext = new ProcessContext(programContext);
        ProgramRunner programRunner = new ProgramRunner(processContext);
        programRunner.run();
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionPosition());
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionsExecuted());
    }

    @Test
    public void testLoop() throws ProgramException {
        ProgramBuilder builder = new ProgramBuilder();
        builder.PUSH(new byte[]{0});
        builder.JUMP_DESTINATION();
        builder.PUSH(new byte[]{1});
        builder.ADD();
        builder.DUPLICATE(0);
        builder.PUSH(new byte[]{1, 0});
        builder.EQUALS();
        builder.IS_ZERO();
        builder.PUSH(BigInteger.valueOf(1L).toByteArray());
        builder.JUMP_IF();
        Program program = builder.build();
        ProgramContext programContext = new ProgramContext(program);
        ProcessContext processContext = new ProcessContext(programContext);
        ProgramRunner programRunner = new ProgramRunner(processContext);
        programRunner.run();
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionPosition());
        Assert.assertEquals(((program.getInstructions().size() - 1) * 256) + 1, programContext.getInstructionsExecuted());
    }

}

File: src/test/java/com/hileco/cortex/instructions/ProgramBuilderTest.kt
package com.hileco.cortex.instructions

import com.hileco.cortex.instructions.ProgramBuilder.FunctionCallConvention.MEMORY_INDEX_AND_ADDRESS
import com.hileco.cortex.instructions.conditions.EQUALS
import com.hileco.cortex.instructions.io.LOAD
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION
import com.hileco.cortex.instructions.jumps.JUMP_IF
import com.hileco.cortex.instructions.stack.PUSH
import com.hileco.cortex.vm.ProgramStoreZone.CALL_DATA
import com.hileco.cortex.vm.concrete.Program
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.VirtualMachine
import org.junit.Assert
import org.junit.Test
import java.math.BigInteger

class ProgramBuilderTest {
    @Test
    fun test() {
        val programBuilder = ProgramBuilder()
        with(programBuilder) {
            jumpDestination("repeat")
            jumpIf(equals(load(CALL_DATA, push(0)), push(123)), "repeat")
        }
        Assert.assertEquals(listOf(
                JUMP_DESTINATION(),
                PUSH(0),
                LOAD(CALL_DATA),
                PUSH(123),
                EQUALS(),
                PUSH(0),
                JUMP_IF()
        ), programBuilder.build())
    }

    @Test
    fun testFunctionCall() {
        val programBuilder = ProgramBuilder()
        with(programBuilder) {
            internalFunctionCall("multiply", {
                push(123)
                push(123)
            })
            internalFunctionCall("multiply", {
                push(123)
                push(123)
            })
            internalFunction("multiply", {
                multiply()
            })
        }
        val program = Program(programBuilder.build())
        val programContext = ProgramContext(program)
        val virtualMachine = VirtualMachine(programContext)
        val programRunner = ProgramRunner(virtualMachine)
        programRunner.run()
        Assert.assertEquals((123 * 123).toBigInteger(), BigInteger(programContext.stack.pop()))
        Assert.assertEquals((123 * 123).toBigInteger(), BigInteger(programContext.stack.pop()))
        Assert.assertTrue(programContext.stack.isEmpty())
    }

    @Test
    fun testNestedFunctionCalls() {
        val programBuilder = ProgramBuilder()
        with(programBuilder) {
            internalFunctionCall("cube", {
                push(123)
            })
            internalFunction("cube", {
                internalFunctionCall("square", {
                    duplicate(1)
                })
                multiply()
            })
            internalFunction("square", {
                duplicate()
                multiply()
            })
        }
        val program = Program(programBuilder.build())
        val programContext = ProgramContext(program)
        val virtualMachine = VirtualMachine(programContext)
        val programRunner = ProgramRunner(virtualMachine)
        programRunner.run()
        Assert.assertEquals((123 * 123 * 123).toBigInteger(), BigInteger(programContext.stack.pop()))
        Assert.assertTrue(programContext.stack.isEmpty())
    }

    @Test
    fun testNestedFunctionCallsThroughMemory() {
        val programBuilder = ProgramBuilder()
        with(programBuilder) {
            configureCallConvention()
            internalFunctionCall("cube", {
                push(123)
            }, callConvention = MEMORY_INDEX_AND_ADDRESS)
            internalFunction("cube", {
                internalFunctionCall("square", {
                    duplicate()
                }, callConvention = MEMORY_INDEX_AND_ADDRESS)
                multiply()
            }, callConvention = MEMORY_INDEX_AND_ADDRESS)
            internalFunction("square", {
                duplicate()
                multiply()
            }, callConvention = MEMORY_INDEX_AND_ADDRESS)
        }
        val program = Program(programBuilder.build())
        val programContext = ProgramContext(program)
        val virtualMachine = VirtualMachine(programContext)
        val programRunner = ProgramRunner(virtualMachine)
        programRunner.run()
        Assert.assertEquals((123 * 123 * 123).toBigInteger(), BigInteger(programContext.stack.pop()))
        Assert.assertTrue(programContext.stack.isEmpty())
    }
}